File: solutions/python/onlineauctionsystem/auction_system.py
from user import User
from auction_listing import AuctionListing
from bid import Bid
from typing import List

class AuctionSystem:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance.users = {}
            cls._instance.auction_listings = {}
        return cls._instance

    @classmethod
    def get_instance(cls):
        return cls()

    def register_user(self, user: User):
        self.users[user.id] = user

    def create_auction_listing(self, listing: AuctionListing):
        self.auction_listings[listing.id] = listing

    def search_auction_listings(self, keyword: str) -> List[AuctionListing]:
        return [listing for listing in self.auction_listings.values() if keyword.lower() in listing.item_name.lower()]

    def place_bid(self, listing_id: str, bid: Bid):
        listing = self.auction_listings.get(listing_id)
        if listing:
            listing.place_bid(bid)

File: solutions/golang/onlineauctionsystem/auction_system.go
package onlineauctionsystem

import (
	"strings"
	"sync"
	"time"
)

type AuctionSystem struct {
	users           map[string]*User
	auctionListings map[string]*AuctionListing
	mu              sync.RWMutex
}

var (
	instance *AuctionSystem
	once     sync.Once
)

func GetInstance() *AuctionSystem {
	once.Do(func() {
		instance = &AuctionSystem{
			users:           make(map[string]*User),
			auctionListings: make(map[string]*AuctionListing),
		}
	})
	return instance
}

func (as *AuctionSystem) RegisterUser(user *User) {
	as.mu.Lock()
	defer as.mu.Unlock()
	as.users[user.ID] = user
}

func (as *AuctionSystem) GetUser(userID string) *User {
	as.mu.RLock()
	defer as.mu.RUnlock()
	return as.users[userID]
}

func (as *AuctionSystem) CreateAuctionListing(listing *AuctionListing) {
	as.mu.Lock()
	as.auctionListings[listing.ID] = listing
	as.mu.Unlock()

	// Start auction timer
	go func() {
		time.Sleep(listing.Duration)
		listing.CloseAuction()
	}()
}

func (as *AuctionSystem) SearchAuctionListings(keyword string) []*AuctionListing {
	as.mu.RLock()
	defer as.mu.RUnlock()

	var matchingListings []*AuctionListing
	keyword = strings.ToLower(keyword)

	for _, listing := range as.auctionListings {
		if strings.Contains(strings.ToLower(listing.ItemName), keyword) ||
			strings.Contains(strings.ToLower(listing.Description), keyword) {
			matchingListings = append(matchingListings, listing)
		}
	}
	return matchingListings
}

func (as *AuctionSystem) PlaceBid(auctionListingID string, bid *Bid) bool {
	as.mu.RLock()
	listing, exists := as.auctionListings[auctionListingID]
	as.mu.RUnlock()

	if exists {
		return listing.PlaceBid(bid)
	}
	return false
}
